onnxruntime/core/platform/check_intel.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "check_intel.h"
+
+#if (defined(_M_AMD64) || defined(__x86_64__))
+#if defined(__linux__)
+#include <cpuid.h>
+#elif defined(_WIN32)
+#include <intrin.h>
+#endif
+#endif
+
+namespace onnxruntime {
+
+CheckIntelResult CheckIntel() {
+  CheckIntelResult intel_check = {false, false};
+  bool is_intel = false;
+  bool is_intel_specified_platform = false;
+
+#if (defined(_M_AMD64) || defined(__x86_64__))
+#if defined(_WIN32)
+  constexpr unsigned int kVendorID_Intel[] = {0x756e6547, 0x6c65746e, 0x49656e69};  // "GenuntelineI"
+  constexpr unsigned int kVendorID_IntelSpecifiedPlatformIDs[] = {
+      // ExtendedModel, ExtendedFamily, Family Code, and Model Number
+      0xa06a,  // MTL
+      0xc065,  // ARL-H
+      0xb065   // ARL-U
+  };
+
+  int regs_leaf0[4];
+  int regs_leaf1[4];
+  __cpuid(regs_leaf0, 0);
+  __cpuid(regs_leaf1, 0x1);
+
+  is_intel =
+      (kVendorID_Intel[0] == static_cast<unsigned int>(regs_leaf0[1])) &&
+      (kVendorID_Intel[1] == static_cast<unsigned int>(regs_leaf0[2])) &&
+      (kVendorID_Intel[2] == static_cast<unsigned int>(regs_leaf0[3]));
+
+  if (!is_intel) {
+    return intel_check;  // if not an Intel CPU, return early
+  }
+
+  for (auto intelSpecifiedPlatform : kVendorID_IntelSpecifiedPlatformIDs) {
+    if (static_cast<unsigned int>(regs_leaf1[0] >> 4) == intelSpecifiedPlatform) {
+      is_intel_specified_platform = true;
+    }
```

Comment:
- nit: variable naming
- perhaps we should use unsigned right shift. signed right shift would fill in leading 1's if the sign bit is set.
- early return?
```suggestion
  for (auto intel_specified_platform : kVendorID_IntelSpecifiedPlatformIDs) {
    if ((static_cast<unsigned int>(regs_leaf1[0]) >> 4) == intel_specified_platform) {
      is_intel_specified_platform = true;
      break;
    }
```